upup/pkg/kutil/rollingupdate_cluster.go

```diff
@@ -269,9 +270,42 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
 	return n
 }
 
-func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval time.Duration, k8sClient *k8s_clientset.Clientset) error {
+func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval time.Duration, headroom int64, k8sClient *k8s_clientset.Clientset) error {
 	c := cloud.(awsup.AWSCloud)
 
+	if headroom > 0 {
+		describeParams := &autoscaling.DescribeAutoScalingGroupsInput{
+			AutoScalingGroupNames: []*string{aws.String(n.ASGName)},
+			MaxRecords: aws.Int64(1),
+		}
+
+		out, err := c.Autoscaling().DescribeAutoScalingGroups(describeParams)
+		if err != nil {
+			return fmt.Errorf("Error adding headroom to ASG %s: %v", n.ASGName, err)
+		}
+
+		currentCapacity := aws.Int64Value(out.AutoScalingGroups[0].DesiredCapacity)
+		desiredCapacity := currentCapacity + headroom
+
+		if desiredCapacity > *out.AutoScalingGroups[0].MaxSize {
+			return fmt.Errorf("Adding headroom exceeds ASG max, increase ASG max to continue.")
```

Comment:
@ese there's a PR open for this https://github.com/kubernetes/contrib/pull/1982 hope to get it in soon. 

@justinsb I didn't know most people use a fixed size ASG 🤔. You definitely raised some good points. 

What did you mean by 
> I also wonder how this interacts with the cluster-autoscaler. Presumably the autoscaler would terminate some pods soon after we added the "headroom" pods.

The headroom is only for nodes so that pods won't go into pending state after nodes are terminated. 

With the current state of cluster autoscaler, I think option 3 might be best. In the near future when cluster-autoscaler is more developed I think it could work for mission critical ASGs as well.  I definitely need to think about this more, I'll post any more thoughts in this PR or any related issues :)